src/pyqasm/subroutines.py

```diff
@@ -146,13 +160,102 @@ def _process_classical_arg_by_value(
                     error_node=fn_call,
                     span=fn_call.span,
                 )
+
+            if isinstance(formal_arg, ExternArgument):
+                formal_arg_type = formal_arg.type
+                formal_arg_size = None
+                if hasattr(formal_arg_type, "size") and formal_arg_type.size is not None:
+                    formal_arg_size, _ = Qasm3ExprEvaluator.evaluate_expression(
+                        formal_arg_type.size
+                    )
+                actual_arg_var = cls.visitor_obj._scope_manager.get_from_global_scope(
+                    actual_arg_name
+                )
+                if (
+                    actual_arg_var.base_type != formal_arg_type
+                    or actual_arg_var.base_size != formal_arg_size
+                ):
+                    if formal_arg_size is not None:
+                        raise_qasm3_error(
+                            f"Argument type mismatch in function '{fn_name}', expected "
+                            f"{type(formal_arg_type).__name__}[{formal_arg_size}] but got "
+                            f"{type(actual_arg_var.base_type).__name__}"
+                            f"[{actual_arg_var.base_size}]",
+                            error_node=fn_call,
+                            span=fn_call.span,
+                        )
+
         actual_arg_value = Qasm3ExprEvaluator.evaluate_expression(actual_arg)[0]
 
-        # save this value to be updated later in scope
+        if isinstance(formal_arg, ExternArgument):
+            # Generate a unique name for the extern argument variable
+            _name = fn_name
+            while not cls.visitor_obj._scope_manager.check_in_scope(_name) and _name == fn_name:
+                _name = f"{fn_name}_{random.randint(1, 1_000_000_000)}"
```

Comment:
better to use the [`uuid`](https://docs.python.org/3/library/uuid.html#uuid.uuid4) instead of a while loop for generating a unique suffix